### PR TITLE
402 record result of candidate on ongoing exam dont wait for exam to end and then calculate result

### PR DESCRIPTION
**Pull request by yash6102000**
if the candidate starts the exam then added ongoing state 
calculate the result after every section

### CODE REVIEW

**Comment by anuragP04** [comment]
Screenshot 2022-12-26 at 10 16 21 AM
When the test is ongoing, we have his name so display it and make it clickable

**Comment by anuragP04** [comment]
Please make all the changes that we have discussed.

File: app/components/barGraph/barGraph.tsx [hunk 1]
```diff
@@ -2,18 +2,22 @@ import Highcharts from 'highcharts'
...
 import HighchartsReact from 'highcharts-react-official'
 import moment from 'moment'
-import type { SectionWiseResults } from '~/interface/Interface'
 
 const BarGraph = ({
-  sectionWiseResult,
+  candidateTestWiseResult,
 }: {
-  sectionWiseResult: Array<SectionWiseResults>
+  candidateTestWiseResult: Array<any>
```
Comment (anuragP04) [inline app/components/barGraph/barGraph.tsx:9]: avoid any..

File: app/routes/$workspaceId/results/groupByTests/$testId.$candidateId.tsx [hunk 2]
```diff
@@ -15,21 +14,17 @@ export const loader: LoaderFunction = async ({ request, params }) => {
...
   invariant(params.testId, 'resultId not found')
-  const candidateResult = await getResultsOfIndividualCandidates({
-    id: params.candidateResultId as string,
-  })
-  const sectionWiseResult = await getSectionWiseResultsOfIndividualCandidate({
-    testId: candidateResult?.testId as string,
-    candidateTestId: candidateResult?.candidateTestId as string,
-  })
-  if (!candidateResult) {
+  const candidateTestWiseResult =
```
Comment (anuragP04) [inline app/routes/$workspaceId/results/groupByTests/$testId.$candidateId.tsx:17]: you are getting section wise result and naming it test wise result...

File: app/components/results/ResultDetails.tsx [hunk 3]
```diff
@@ -3,32 +3,40 @@ import { Icon } from '@iconify/react'
...
 import EmptyStateComponent from '../common-components/EmptyStateComponent'
+import type { SectionInCandidateTest } from '~/interface/Interface'
 // import { useTranslation } from 'react-i18next'
 
 const ResultDetailsComponent = () => {
   // const { t } = useTranslation()
 
-  const { candidateResult, params, sectionWiseResult } = useLoaderData()
-
+  const { params, candidateTestWiseResult, currentWorkspaceId } =
```
Comment (anuragP04) [inline app/components/results/ResultDetails.tsx:14]: candidate testWiseResult can be candidateTestResult

File: app/components/results/ResultDetails.tsx [hunk 4]
```diff
@@ -38,30 +46,35 @@ const ResultDetailsComponent = () => {
...
         <>
-          <BarGraph sectionWiseResult={sectionWiseResult} />
+          <BarGraph candidateTestWiseResult={candidateTestWiseResult} />
           <Divider height="1px" />
           <div
             id="results-test-candidate-list-tab"
             className="flex flex-col gap-6"
           >
-            {sectionWiseResult.map((section: SectionWiseResults) => {
+            {sortSections.reverse().map((section: SectionInCandidateTest) => {
```
Comment (anuragP04) [inline app/components/results/ResultDetails.tsx:64]: instad of using reverse, we can have sort in reverse order

File: app/components/barGraph/barGraph.tsx [hunk 5]
```diff
@@ -2,61 +2,65 @@ import Highcharts from 'highcharts'
...
+import type {
+  CandidateTest,
+  SectionInCandidateTest,
+  SectionInTest,
+  SectionWiseResults,
+} from '~/interface/Interface'
 
 const BarGraph = ({
-  sectionWiseResult,
+  candidateTestWiseResult,
```
Comment (anuragP04) [inline app/components/barGraph/barGraph.tsx:13]: it can be candidateTestResult

File: app/components/barGraph/barGraph.tsx [hunk 6]
```diff
@@ -2,61 +2,65 @@ import Highcharts from 'highcharts'
...
+  calculateResult.map((data: SectionInCandidateTest) => {
+    return data.SectionWiseResult?.map((results: SectionWiseResults) => {
+      let data = results.section.section.sectionInTest.map(
+        (data: SectionInTest) => {
+          return data
+        }
+      )
+      result.push(Math.floor(data[0].timeInSeconds / 60))
+    })
+  })
```
Comment (anuragP04) [inline app/components/barGraph/barGraph.tsx:45]: can you add comment what these lines are doing

File: app/components/barGraph/barGraph.tsx [hunk 7]
```diff
@@ -2,61 +2,65 @@ import Highcharts from 'highcharts'
...
-      return getRequiredSection?.totalQuestion
+      return getRequiredSection[0]?.SectionWiseResult[0]?.totalQuestion
     } else if (resultKind === 'correct') {
-      return getRequiredSection?.correctQuestion
+      return getRequiredSection[0]?.SectionWiseResult[0]?.correctQuestion
     } else if (resultKind === 'skipped') {
       return (
-        (getRequiredSection?.totalQuestion as number) -
-        (getRequiredSection?.correctQuestion as number)
+        (getRequiredSection[0]?.SectionWiseResult[0]?.totalQuestion as number) -
```
Comment (anuragP04) [inline app/components/barGraph/barGraph.tsx:59]: why we are using hardcoded index here?

File: app/models/assessment.server.ts [hunk 8]
```diff
@@ -586,8 +591,134 @@ export async function endAssessment(id: string) {
     await sendMailToRecruiter(recruiterEmail, testName, candidateName)
   return endExam
 }
+async function calculateResultByIdSection(sectionid?: string) {
```
Comment (anuragP04) [inline app/models/assessment.server.ts:594]: does this name make any sense??

File: app/components/barGraph/barGraph.tsx (hunk 6 above)
Comment (yash6102000) [reply app/components/barGraph/barGraph.tsx:45]: discussed

File: app/components/barGraph/barGraph.tsx [hunk 9]
```diff
@@ -2,61 +2,65 @@ import Highcharts from 'highcharts'
...
-        sectionWiseResult[j].section.section.id ===
-        getSectionsFromResult[j][k].section.id
-      ) {
-        result.push(Math.floor(getSectionsFromResult[j][k].timeInSeconds / 60))
-      }
-    }
-  }
+  calculateResult.map((data: SectionInCandidateTest) => {
+    return data.SectionWiseResult?.map((results: SectionWiseResults) => {
+      let data = results.section.section.sectionInTest.map(
```
Comment (richa-copods) [inline app/components/barGraph/barGraph.tsx:38]: Use `const` and rename

File: app/components/results/AttendedCandidateListItem.tsx [hunk 10]
```diff
@@ -115,23 +113,29 @@ const AttendedCandidateListItem = ({
...
+              {t('commonConstants.onGoing')}
+            </span>
+          ) : endAt != null ? (
+            <span className="rounded-full bg-green-200 px-2 py-1 text-xs">
+              {t('commonConstants.complete')}
+            </span>
+          ) : (
+            ''
+          )}
           {result >= 0 ? (
```
Comment (richa-copods) [inline app/components/results/AttendedCandidateListItem.tsx:139]: Change the condition to `result<0`

File: app/models/assessment.server.ts [hunk 11]
```diff
@@ -586,8 +591,134 @@ export async function endAssessment(id: string) {
...
+      ? section?.questions?.length
+      : 0
+    let unanswered = 0
+    let correct = 0
+    let incorrect = 0
+    let skipped = 0
+    for (let question of section?.questions || []) {
+      // counting unanswered questions
+      if (
+        question?.answers.length == 0 &&
```
Comment (richa-copods) [inline app/models/assessment.server.ts:634]: Use strict check `===`. For entire file

**Review by richa-copods — CHANGES_REQUESTED** [review]

**Review by anuragP04 — APPROVED** [review]

**Pull request merged**